articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
 
-There are some cases where we can't pass a parameter into the procedure call, or a query, and string appending is the only option.
+== Role Based Privileges
 
-For example, there are some things in Cypher that cannot be parameterized, such as node labels and relation types.
-There are some APOC Procs that can help (and should be used if so), but aside from these, the only option may be to append into a query string.
+=== Principle of Least Privilege
 
-In these cases, it is extremely important to sanitize your inputs, removing quote or delimiter characters (depending on their context of use) that would allow input to break out of the context within which you're trying to use it.
+The principle of least privilege is the idea that a program or user should have the bare minimum of privileges needed to perform their function.
+This means that, for example, if your application is only reading data, then it should have read only access to that data.
+The benefit of this is that in the case of a Cypher Injection attack, the injected query is not able to manipulate the data,
+as the role executing the hijacked query is limited in its functionality.
+With Neo4j, a range of fine-grained access control is available, which can add another layer of protection in case of an injection attack.
+See https://neo4j.com/docs/operations-manual/current/authentication-authorization/access-control/[here] for more information on role based privileges in Neo4j.
 
-In these cases it is better to sanitize input in your own code at the client level, as there are many utilities across various languages for input sanitization, and it makes sense to address it at that level rather than lower down at the database itself.
+== Other Notes
+
+Not all inputs can be submitted as parameters.
+Maybe some malicious input made it into a CSV file for processing.
+A CSV of the names of new students for the year, for example.
+
+[source,cypher]
+----
+LOAD CSV WITH HEADERS FROM "file:///students_2021.csv" AS row
+CREATE (s:Student)
+SET s.year = 2021, s.name = row.student_name
+----
+
+Is this vulnerable to Little Robby Labels?
+
+No, it is not. Cypher injection is still impossible here, even if parameters aren't being used.
+
+The LOAD query is independent of the CSV that is to be processed.
+As such, the query is compiled separately from the CSV.
+By the time the query is executing and the CSV data starts to be accessed, the query has already been completely compiled, and the CSV data has no opportunity to affect or hijack the query itself.
+
+Likewise, any other injection attacks will fail that attempt to pull in malicious input via reading it from somewhere, or even a malicious property value that somehow was saved into the database already.
+This is because unless the malicious value is string-appended into the query itself, rather than read at the time the query executes, it will not have the opportunity to get compiled into and affect the query plan.
```

Comment:
This section could perhaps be worked into a little higher in the document. At some point we mention what needs to be parameterised/sanitised, maybe we could also have a section about what doesn't need to be.

Reply:
_Food for thought._

Alternatively, although it's good information, we could delete this section. 

It's one thing to document what needs to be sanitised, and it's another to begin discussing what doesn't need to be. You can't really enumerate all the things which don't need to be sanitised, in the same way that you can't proving a negative.

Reply:
Given the frequency of load and import queries, I think it's important to continue to include this section, and show that malicious input in the source files aren't vulnerable.